src/neo/SmartContract/InteropService.Storage.cs

```diff
@@ -24,9 +24,34 @@ public static class Storage
             public static readonly InteropDescriptor PutEx = Register("System.Storage.PutEx", Storage_PutEx, GetStoragePrice, TriggerType.Application, CallFlags.AllowModifyStates);
             public static readonly InteropDescriptor Delete = Register("System.Storage.Delete", Storage_Delete, 0_01000000, TriggerType.Application, CallFlags.AllowModifyStates);
 
-            private static long GetStoragePrice(EvaluationStack stack)
+            private static long GetStoragePrice(ApplicationEngine engine)
             {
-                return (stack.Peek(1).GetByteLength() + stack.Peek(2).GetByteLength()) * GasPerByte;
+                var stack = engine.CurrentContext.EvaluationStack;
+                var key = stack.Peek(1);
+                var value = stack.Peek(2);
+                var newDataSize = value.IsNull ? 0 : value.GetByteLength();
+                if (!(engine.CurrentContext.EvaluationStack.Peek() is InteropInterface _interface))
+                    throw new InvalidOperationException();
+
+                StorageContext context = _interface.GetInterface<StorageContext>();
+                StorageKey skey = new StorageKey
+                {
+                    Id = context.Id,
+                    Key = key.GetSpan().ToArray()
+                };
+                var skeyValue = engine.Snapshot.Storages.TryGet(skey);
+                if (skeyValue is null || skeyValue.Value is null || skeyValue.Value.Length == 0)
+                    return (key.GetByteLength() + newDataSize) * GasPerByte;
+
+                var currentOccupiedBytes = skeyValue.Value.Length;
+                if (newDataSize <= currentOccupiedBytes)
+                {
+                    return 1 * GasPerByte;
+                }
+                else
+                {
+                    return (newDataSize - currentOccupiedBytes) * GasPerByte;
```

Comment:
We want to allow negative? if in the future we have free tx, we could use free tx for increase byte per byte, and reduce with other tx, have negative fee, and create a `while true`.

Reply:
As discussed at #1452 , free tx only handle NEO transfer and voting. Don't need to consider negative fee?  Or we can fix it in the future when that happen?

Reply:
I think we should not allow negative fee in any case.